File: py/1888.py
class Solution:
    def minFlips(self, s: str) -> int:
        s2 = s * 2
        alt1 = ''.join([str(i % 2) for i in range(len(s2))])
        alt2 = ''.join([str((i + 1) % 2) for i in range(len(s2))])

        diff1 = 0
        diff2 = 0

        res = float('inf')

        l = 0
        for r in range(len(s2)):
            # 先更新窗口
            if s2[r] != alt1[r]:
                diff1 += 1
            if s2[r] != alt2[r]:
                diff2 += 1
            # 当r == len(s) - 1时就可以取值了
            if r < len(s) - 1:
                continue
            res = min(res, diff1, diff2)

            if s2[l] != alt1[l]:
                diff1 -= 1
            if s2[l] != alt2[l]:
                diff2 -= 1

            l += 1

        return res


ss = ["111000", "010", "1110"]

for s in ss:
    print(Solution().minFlips(s))

File: go/1888.go
package main

import "math"

func minFlips(s string) int {
	res := math.MaxInt

	ds := s + s
	alt1 := make([]byte, len(ds))
	alt2 := make([]byte, len(ds))

	for i := range ds {
		alt1[i] = '0' + byte(i%2)
		alt2[i] = '0' + byte((i+1)%2)
	}

	diff1 := 0
	diff2 := 0

	l := 0
	for r := range ds {
		if alt1[r] != ds[r] {
			diff1 += 1
		}
		if alt2[r] != ds[r] {
			diff2 += 1
		}
		if r < len(s)-1 {
			continue
		}
		if r-l+1 == len(s) {
			res = min(res, diff1)
			res = min(res, diff2)
		}
		if alt1[l] != ds[l] {
			diff1 -= 1
		}
		if alt2[l] != ds[l] {
			diff2 -= 1
		}
		l++
	}

	if res == math.MaxInt {
		return 0
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}